User can log in

piavW opened the pull request:
[PT-link](https://www.pivotaltracker.com/story/show/169253769)

## PR content
Allows a user to log in to their account on the application.

## This PR contains
- Redux-token-auth package and configuration
- Login functionality
- POST-request towards back-end. 

## What we've learned
- Always use prevent-default with form-tags (if using form-tags add url and don't use buttons - use input-tag-submit instead)
- Improved knowledge of debugging and Redux-Token-Auth

<img width="963" alt="Screenshot 2019-10-28 at 13 17 56" src="https://user-images.githubusercontent.com/53825385/67677884-7c20c600-f985-11e9-9f20-e9574ddc66e9.png">
<img width="965" alt="Screenshot 2019-10-28 at 13 18 09" src="https://user-images.githubusercontent.com/53825385/67677885-7c20c600-f985-11e9-9ded-d571aa96cae8.png">

piavW committed:
Creates integration-test.

jonas-masharqa committed:
Creates LoginForm component,
- adds email and password states in app
- creates variable for rendering login button or form

piavW committed:
Creates fixture file, updates login-form.

piavW committed:
Installs and configures redux-token-auth, wrapps app wihin redux.

jonas-masharqa committed:
Imports signInUser and connect to App and exports currentUsers state and mapDispatchToProps.

piavW committed:
[WIP] Login-functionality with redux
- Refactors integration test
- Adds getArticlesIndex state in Apps,
- Commented out componentDidMount
- Creates state-changing functions
- Creates handleLogin with ReduxTokenAuth actions/attributes
- Adds welcome-message
- Inputs debuggers

jonas-masharqa committed:
Comments out getArticesData function and componentDidMount, continous debugging.

piavW committed:
Cleans up code back in initial state before debugging, removes unnecessary debuggers. Fixes spelling error of fixtures in integration-test.

piavW committed:
Remove form-tag and replaces with div in login-form-component. Removes debuggers.

clalil commented on src/App.js:
```diff
@@ -1,15 +1,94 @@
+import { connect } from 'react-redux'
 
 class App extends Component {
+  state = {
```
Is it possible to extract the entire Login functionality into a separate component as to avoid cluttering our App component? With redux, the state should be OK outside of the App component, right?
i.e.
class App...
return (
_Login component_
)

class Login...
return (
_LoginForm component_
)

piavW replied:
We feel it's okay to have it in the App.js since we've used it before and we'd still need to send in data into that new component. It's not priority to refactor the code to that extent at the moment. But it is something we can revisit later if time allows.

clalil replied:
Since I am a minority on this case, I will accept. However, I am concerned about this becoming an issue later down the line f.eks. let's say the sign in functionality would require as much code as the login functionality, and you put all of it inside of the app component - would you still think it was easy to follow and track possible bugs? As opposed to knowing all of the code for the login-functionality is all extracted and restricted to one single component.

piavW replied:
We're thinking we'll cross that bridge when we get there :) But we agree it might be an issue later down the line

clalil commented on src/App.js:
```diff
@@ -1,15 +1,94 @@
+    getArticlesIndex: true
+  }
+
+  renderArticleGetState = () => {
```
We do not need this anymore, right? Please remove.

clalil replied:
Why would you need a closing tag that has no corresponding opening tag?
```
        <**Input** iconPosition='left' placeholder="Email" id="email-input" name="email" onChange={props.inputChangeHandlerProps}>
          <Icon name='at' />
          <input />
        **</Input>**
        **<Input** id="password-input" placeholder="Password" name="password" onChange={props.inputChangeHandlerProps}>
          <input />
        **</Input>**
```

piavW replied:
Because if you remove the <input /> then there is not input field on the page. I've tired it out on the semantic-ui-react website.

clalil replied:
That is really strange...on the cooper challenge, why wasn't that an issue?

```
          <Input
            label={{ basic: true, content: 'Years' }}
            labelPosition='right'
            placeholder='Enter your age...'
            id="age" onChange={props.inputChangeHandler}
          />
```

clalil commented on src/App.js:
```diff
@@ -1,15 +1,94 @@
+        {renderLogin}
+        {welcomeMessage}
+
+        <ListArticles 
```
Don't need this anymore, right? Please revert back to upstream development state for this component

clalil replied:
Same as above.

clalil commented on src/Components/ListArticles.js:
```diff
@@ -28,10 +34,14 @@ class ListArticles extends Component {
     let articleList
     let errorMessage
 
+    if (this.props.getArticlesIndexProps === true) {
```
Don't need this anymore, right? Please revert component back to upstream development branch state.

clalil commented on src/Components/LoginForm.js:
```diff
@@ -0,0 +1,17 @@
+import React, { Component } from 'react'
```
Minor detail, but this component does not need to be a class component. Could we make it into a functional component as to avoid using class components when they are not needed?

clalil replied:
Also, are we happy with the styling of this component. I.e. could we perhaps add a button UI component?

jonas-masharqa committed:
Removes irrelevant getArticleIndex state from app and functionality.

jonas-masharqa committed:
Configures LoginForm component to be functional instead of class.

piavW committed:
Adds styling to login-form, buttons and welcome message.

clalil commented on src/Components/LoginForm.js:
```diff
@@ -0,0 +1,21 @@
+          <input />
+        </Input>
+        <Input id="password-input" placeholder="Password" name="password" onChange={props.inputChangeHandlerProps}>
+          <input />
```
Remove remaining additional closing tag.

piavW replied:
No we need that tag because otherwise there is no input field.

clalil replied:
Same as above.

clalil replied:
Why would you need a closing tag that has no corresponding opening tag?
```
        <**Input** iconPosition='left' placeholder="Email" id="email-input" name="email" onChange={props.inputChangeHandlerProps}>
          <Icon name='at' />
          <input />
        **</Input>**
        **<Input** id="password-input" placeholder="Password" name="password" onChange={props.inputChangeHandlerProps}>
          <input />
        **</Input>**
```

clalil commented on src/Components/LoginForm.js:
```diff
@@ -0,0 +1,21 @@
+      <Container id="login-form">
+        <Input iconPosition='left' placeholder="Email" id="email-input" name="email" onChange={props.inputChangeHandlerProps}>
+          <Icon name='at' />
+          <input />
```
Remove remaining additional closing tag.

piavW replied:
Same as above - needed because of how semanic-ui-react works.

clalil replied:
Why would you need a closing tag that has no corresponding opening tag?
```
        <**Input** iconPosition='left' placeholder="Email" id="email-input" name="email" onChange={props.inputChangeHandlerProps}>
          <Icon name='at' />
          <input />
        **</Input>**
        **<Input** id="password-input" placeholder="Password" name="password" onChange={props.inputChangeHandlerProps}>
          <input />
        **</Input>**
```

oliverochman reviewed (commented):
Please add sad paths for what happens when user fills in invalid credentials

piavW committed:
Adds sad-path integration tests.

piavW committed:
Continues writing sad-path
- Updates integration-test's sadpath
- Creates sadpath fixture-file
- Creates loginData module to POST LoginData
- Adds responseMessage for error-message in App.js

jonas-masharqa committed:
Calls on loginDataHandler in handleLogin, adds debuggers, adds it block in integration test.

piavW committed:
Creates Login component
- Creates new Login component
- Refactors App and moves login-functionality to the Login component

piavW committed:
Cleans up app.js

piavW commented:
Closing due to team-member changes.
